client/question/question.js

```diff
@@ -550,101 +550,195 @@ tie.constant(
+        var re = new RegExp('^NameError: global name .* is not defined$');
+        return re.test(errorString);
+      },
+      getFriendlyErrorText: function(errorString) {
+        var re = new RegExp('NameError: global name (.*) is not defined');
+        var varName = errorString.match(re)[1];
+        return [
+          '<code>', varName, '</code> appears to be a name that was not ',
+          'previously declared or defined. E.g., if <code>', varName,
+          '</code> is a variable to hold a string, <code>', varName,
+          ' = \'\'</code> would declare <code>', varName, '</code> as a ',
+          'string variable. Similarly, <code>', varName, ' = []</code> ',
+          'declares <code>', varName, '</code> as an array variable. Another ',
+          'possibility is that <code>', varName, '</code> was misspelled or ',
+          'uses incorrect capitalization.'].join('');
+      }
+    }, {
+      friendlyErrorCheck: function(errorString) {
+        var re = new RegExp('^SyntaxError: invalid syntax');
+        return re.test(errorString);
+      },
+      getFriendlyErrorText: function() {
+        return [
+          'This is the default, generic syntax error message in Python. Here ',
+          'are some common mistakes that can result in this error:',
+          '<ul>',
+          '<li>Mis-matching (missing or too many) braces, brackets, ',
+          'parentheses, or quotation marks',
+          '<li>Missing colons when defining a function (e.g., ',
+          '<code>def my_function</code> should be ',
+          '<code>def my_function:</code>)',
+          '<li>Misspelled Python keywords (e.g., misspelling ',
+          '<code>for</code> in a for loop)',
+          '</ul>',
+          'Note that since this is the default, generic error, it could also ',
+          'be caused by something else.'].join('');
+      }
+    }, {
+      friendlyErrorCheck: function(errorString) {
+        var re = new RegExp('^SyntaxError: EOL while scanning string literal');
+        return re.test(errorString);
+      },
+      getFriendlyErrorText: function() {
+        return [
+          'An "End of Line" error on a string usually means you are ',
+          'missing a quotation mark somewhere.'].join('');
+      }
+    }, {
+      friendlyErrorCheck: function(errorString) {
+        var re = new RegExp('^IndentationError: ');
+        return re.test(errorString);
+      },
+      getFriendlyErrorText: function() {
+        return [
+          'A line of code might not be indented properly. Python is ',
+          'particular about indentation, since that\'s how it identifies ',
+          'code blocks.'].join('');
+      }
+    }, {
+      friendlyErrorCheck: function(errorString) {
+        var re = new RegExp('^TypeError: can only concatenate');
+        return re.test(errorString);
+      },
+      getFriendlyErrorText: function() {
+        return [
+          'You may have tried to combine two incompatible types (you might ',
```

Comment:
Optional: Perhaps add "e.g." at the beginning of the parenthetical phrase, and strike "etc." at the end. (I think this is trying to convey "for example" rather than "and others".)

Reply:
Hmm... Yeah, I originally had "e.g.", but I wanted to emphasize that there are other examples. The concern being that the novice would focus too much on the example given.

Reply:
Hmm, I see what you mean. I'm fairly neutral on this one; your call.